rx/concurrency/virtualtimescheduler.py

```diff
@@ -12,129 +14,155 @@
 
 MAX_SPINNING = 100
 
+
 class VirtualTimeScheduler(SchedulerBase):
     """Virtual Scheduler. This scheduler should work with either
     datetime/timespan or ticks as int/int"""
 
     def __init__(self, initial_clock=0.0) -> None:
         """Creates a new virtual time scheduler with the specified
-        initial clock value and absolute time comparer.
+        initial clock value.
 
         Args:
             initial_clock: Initial value for the clock.
-            comparer: Comparer to determine causality of events based
-                on absolute time.
         """
-        self.clock = initial_clock
-
-        self.is_enabled = False
-        self.queue = PriorityQueue(1024)
+        self._clock = initial_clock
```

Comment:
Maybe I'm missing something, but making the `clock` member pseudo-private (by convention) seems to be a bit misleading. It's explicitly used elsewhere, e.g. by the `HistoricalScheduler` or in tests. What do you think?

Reply:
Ah, good point! I made it private because setting this clock from unsynchronized code will cause problems, and figured it would be alright for “trusted” code like HistoricalScheduler and tests to access it.

But you’re right, of course, it would have been better to add a public read-only property here. I will do that tonight and push to this PR.

Thanks!!

Reply:
:+1: I will still approve this PR so you should be able to merge, since I may be not available tomorrow. Great Job!

Reply:
@jcafhe I just pushed a change that addresses your point. Does that look all right? Thanks again for the thorough review, this is a big PR and it makes me feel slightly more secure knowing that multiple eyes went over it!

Reply:
Hmm, sorry, I've had to force-push that last bit twice -- I had a little change in my local working copy and that got included by accident. Fixed now.

Reply:
yes that feels good for me :+1: 
> this is a big PR and it makes me feel slightly more secure knowing that multiple eyes went over it!

I must admit I've struggled with the run loop in `EventLoopScheduler`, I didn't manage to fully understand what's going on.

Reply:
Glad that you mention that, it means the code change isn't self-explanatory and so I should have been more verbose about the changes in EventLoopScheduler, besides the synchronization of its queue.

Currently, waiting until the up-next item is due looks like this:
- EventLoopScheduler starts a separate Timer and blocks, waiting for its threading.Condition to be notified. 
- The timer is just a secondary thread which has a threading.Event on which it waits with a timeout, unless the event is set (cancelled).
- After timeout, this secondary thread then invokes a callback, which in this case is the tick method of our scheduler.
- Finally, the tick method notifies the original threading.Condition and our main loop resumes.

So it occurred to me that the main loop, instead of doing an indefinite wait for an outside timer, ought to simply add a timeout argument to its own initial wait.

Furthermore, while considering this code, it seemed to me that it was possible for edge cases to go out-of-order, due to the fact that there are two routes for an item to being invoked (queue and ready_list). So I added a little bit of logic that makes sure those are merged / zipped correctly into each iteration's actual work order.

Does this help?

Reply:
Thank you for your deep insight, it is very helpful. I understand now.
Just to be sure, I understand that there's a sort by duetime between items from `_queue` and items from the `_ready_list` before appending to the `ready` deque. To be more accurate, the comparison is done with the first item in `_ready_list` so it is assumed that `_ready_list` is already sorted in some way.

Reply:
Forget my comment, `_ready_list` should be de facto sorted since it's only used to schedule ASAP.

Reply:
Yes, that’s exactly right; we can do a merge here rather than a full sort, since ready_list and the queue are each already sorted.

So, do you think the PR is ready to be merged?

Reply:
Yes sure.

Reply:
Excellent, thanks again for the detailed review!